app/scripts/lib/nonce-tracker.js

```diff
@@ -83,12 +78,77 @@ class NonceTracker {
     return mutex
   }
 
+  async _getNetworkNonceAndDetails (address) {
+    // calculate next nonce
+    // we need to make sure our base count
+    // and pending count are from the same block
+    const currentBlock = await this._getCurrentBlock()
+    const blockNumber = currentBlock.blockNumber
+    const baseCountHex = await this._getTxCount(address, currentBlock)
+    const baseCount = parseInt(baseCountHex, 16)
+    assert(Number.isInteger(baseCount), `nonce-tracker - baseCount is not an integer - got: (${typeof baseCount}) "${baseCount}"`)
+    // if the nonce provided by the network is higher then a pending tx
+    // toss out the pending txCount
+    const networkNonce = baseCount
+
+    return {networkNonce, blockNumber, baseCountHex, baseCount}
+  }
+
+  _getLocalNextNonce (address) {
+    const confirmedTransactions = this._reduceTxListToUniqueNonces(this.getConfirmedTransactions(address))
+    const pendingTransactions = this._reduceTxListToUniqueNonces(this.getPendingTransactions(address))
+    const transactions = this._reduceTxListToUniqueNonces(confirmedTransactions.concat(pendingTransactions))
+    let localNonce = this._getHighestNonce(transactions)
+    // throw out localNonce if not a number
+    if (!Number.isInteger(localNonce)) localNonce = 0
+    if (
+      // the local nonce is not 0
+      localNonce ||
+      // or their are pending or confirmed transactions
+      this._getPendingTransactionCount(address) ||
+      confirmedTransactions.length
+    ) ++localNonce
+    return localNonce
+  }
+
+  _getLocalPendingNonce (address) {
+    const pendingTransactions = this._reduceTxListToUniqueNonces(this.getPendingTransactions(address))
+    const localNonce = this._getHighestNonce(pendingTransactions)
+    return localNonce
+  }
+
+  _getPendingTransactionCount (address) {
+    const pendingTransactions = this.getPendingTransactions(address)
+    return this._reduceTxListToUniqueNonces(pendingTransactions).length
+  }
+
+
+  _reduceTxListToUniqueNonces (txList) {
+    const reducedTxList = txList.reduce((reducedList, txMeta, index) => {
+      if (!index) return [txMeta]
+      const nonceMatches = txList.filter((txData) => {
+        return txMeta.txParams.nonce === txData.txParams.nonce
+      })
+      if (nonceMatches.length > 1) return reducedList
+      reducedList.push(txMeta)
+      return reducedList
+    }, [])
+    return reducedTxList
+  }
+
+  _getHighestNonce (txList) {
+    const nonces = txList.map((txMeta) => txMeta.txParams.nonce)
```

Comment:
should move `parseInt` to inside the mapping then you dont need to do it everywhere